c10/util/OptionalArrayRef.h

```diff
@@ -0,0 +1,354 @@
+// Optimization for ArrayRef<T>. We take advantage of an unused bit pattern in
+// ArrayRef (inspired by Arthur O'Dwyer's tombstone_traits -- see
+// https://youtu.be/MWBfmmg8-Yo?t=2466) to keep the size of
+// c10::optional::ArrayRef<T> down to 16 bytes, which allows it to be passed to
+// functions in registers instead of getting passed in memory per item 5c of the
+// classification algorithm in section 3.2.3 of the System V ABI document
+// (https://www.uclibc.org/docs/psABI-x86_64.pdf).
+//
+// c10::optional<ArrayRef<T>> as well as std::optional<ArrayRef<T>> have a
+// problem where the implicit converting constructor will cause the underlying
+// ArrayRef<T> to store a dangling pointer. OptionalArrayRef<T> prevents this.
+// See https://github.com/pytorch/pytorch/issues/63645 for more on this.
```

Comment:
Must we copy paste the implementation here? While it's true that this code is unlikely to change, it would be really nice if we could only have one implementation of the code. What I am imagining is that OptionalArrayRef contains an optional<ArrayRef<T>> and it just replicates the public API from there + fixed constructors.

Reply:
Yeah that makes sense, I'll go ahead and do that

Reply:
Alright, I've simplified it. Should I write some C++ tests for `OptionalArrayRef`, or do you think it's good enough if it just works in all the existing CI? I suppose we definitely should at least have one test to exercise the dangling pointer fix, so I'll add that

Reply:
Just added the one test

Reply:
Just one test sounds good